ggml-cuda.cu

```diff
@@ -5059,6 +5240,173 @@ template <bool need_check> static __global__ void
 #endif // __CUDA_ARCH__ >= CC_VOLTA
 }
 
+#define MMI8_PADDING_X 4
+#define MMI8_PADDING_Y 4
+#define MMI8_COPY_SIZE 2
+
+#define      MMI8_X_AMPERE 64
+#define      MMI8_Y_AMPERE 144
+#define MMI8_NWARPS_AMPERE 4
+
+template <int mmi8_x, int mmi8_y, bool need_check_x, bool need_check_y>
+#if __CUDA_ARCH__ >= CC_AMPERE
+// __launch_bounds__(WARP_SIZE*mmi8_x/16, mmi8_x >= 128 ? 1 : 2)
+__launch_bounds__(WARP_SIZE*mmi8_y/32, 1)
+#else
+__launch_bounds__(WARP_SIZE*MMI8_NWARPS_AMPERE, 2)
+#endif // __CUDA_ARCH__ >= CC_AMPERE
+static __global__ void mul_mat_i8(
+    const int * __restrict__ x_qs_low, const float * x_d, const int * __restrict__ y_qs_low, const int * __restrict__ y_bs,
+    const float * y_d, float * __restrict__ dst, const int ncols_x, const int nrows_x, const int ncols_y,
+    const int nrows_y, const int nrows_dst) {
+
+// #if __CUDA_ARCH__ >= CC_VOLTA && !(defined(GGML_USE_HIPBLAS) && defined(__HIP_PLATFORM_AMD__))
+    typedef nvcuda::wmma::fragment<nvcuda::wmma::matrix_a,    32, 8, 16, int8_t, nvcuda::wmma::row_major> frag_thin_a;
```

Comment:
Re. AMD RDNA3 support - only 16x16x16 sizes are supported.

Reply:
That's a problem. I'm specifically using 32x8 fragments because that way I can define more integer scales for the b fragments so the precision is better.

Reply:
Hmm. Could 32x16 work?

Reply:
I previously had an implementation with 16x16 fragments. It does work but the precision is worse. The increase in perplexity with single precision MMI8 compared to cuBLAS FP16 GEMM was ~40% higher.

Reply:
Noob question - so the higher perplexity is because 16 scales don't provide enough precision compared to 32 scales? if so, then I wonder if 32x16 is worth a try. But I'm unsure of how it would affect the perplexity score.

Reply:
The trick to making the current single-precision implementation work is to calculate additional scales for the "b" fragments, i.e. the fragments of the hidden state. This makes it so that large values in the hidden state only affect the precision of a few values rather than all of the values in a column. So it is better to use small "b" fragments because you can scale the values with a finer granularity. 32x16 and 16x8 is better than 16x16 and 16x16 because the b fragments are smaller (the a fragment size does not matter). So 32x16 would in fact be worse than 16x16.